o3de/o3de
# Remote gem version support

## What does this PR do?

Extends gem version support to remote gems and to some extent, projects and templates.

### Important changes
- do not change existing `gems`, `projects`, `templates` fields in `repo.json` this would break backwards compatibility, instead use new fields to store object json data in `repo.json`  (new fields are `gems_data`, `projects_data`, `templates_data`)
- Gem Catalog now defaults to showing compatible gems.  Compatibility filtering is based on `compatible_engines`, `gem_dependencies` and `engine_api_dependecies` and not based on full dependency resolution which would be too slow.  However, full dependency resolution is performed when a user attempts to save project changes.
- Remote gems are downloaded into versioned subfolders e.g.  `c:/path/to/<gem name>/<gem version>/`  If a gem has no version then `0.0.0` is used.
- https://canonical.o3de.org remote repo is added in `engine.json`
- Remote repos can be activated/deactivated so their content is no longer accessible in the Gem Catalog, this will be necessary for recommended repos

### Other changes/fixes
- added ability to refresh a repo from the o3de CLI using `o3de.bat repo --refresh-repo <repo url>` or all repos `o3de.bat repo --refresh-all-repos`
- added ability to clone a remote object at a specific reference.  If the object data contains a `source_control_uri` and `source_control_ref`.  This is only supported by the CLI currently using `o3de.bat download <object name and optional version specifier> --use-source-control`  e.g. `o3de.bat download XR==1.2.3  --use-source-control`
- add 'Missing' filter and show missing gems versions a project uses
- add Official badge/banner for o3de canonical repo and sort it to top.
![image](https://user-images.githubusercontent.com/26804013/232544428-7ef08b5a-75d4-4519-9897-9a0c2230e205.png)

Screenshot of main gems tab that shows compatible filter and a remote gem available for download (PopcornFX)
![image](https://user-images.githubusercontent.com/26804013/232544818-cbba31e3-2b68-46a4-b878-c2f0d8fbf316.png)

Screenshot showing `Missing` filter and missing local and remote gem versions
![image](https://user-images.githubusercontent.com/26804013/232638516-fcead57b-d06b-4d09-9428-89d193158a00.png)



Relevant RFC: https://github.com/o3de/sig-core/issues/44

## How was this PR tested?

Unit testing:
- updated and created unit test for new functionality, ran all o3de pytests

Manual testing:
- set up a repository and verified I could see compatible, and incompatible content
- downloaded and uninstalled remote gem versions
- verified new versions of PopcornFX could be downloaded and uninstalled
- verified refreshing remote repositories parses and loads new data and last updated field is correctly shown

# Exception Checklist
- [x] 1. Owning SIG requests the exception, by a Sig Representative (Chair/co-chair) adding the PR to the exception requests queue board.
- [x] 2. (SIG-Content) Owning SIG comments on the PR explaining impact if this code does not go into stabilization.
- [x] 3. Unrelated SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] 4. UX SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] 5. Testing SIG representative (chair/cochair) approves the exception, adding a comment indicating approval and the SIG they represent.
 (Must occur at some point before SIG-Release approval) Code has been approved for merging by maintainers
- [x] 6. SIG-Release Release Manager or co-Release Manager approves the exception, adding a comment indicating approval and the SIG they represent.
- [x] 7. If all approvals are accepted, then the exception PR may be merged
- [ ] 8. Merged PR is verified successful via testing. If Merged PR verification testing failed, consult reps from step 3-5 and discussion occurs on fixing the failure vs backing out change
- [ ] 9. Exception is complete
